6.2-c : printing info on subproblem into structure.txt

guilpier-code opened the pull request.

Commit:
6.2-c : rename modeler.solve() into modeler.run()

Commit:
6.2-c : putting things away at the top modeler level

Commit:
6.2-c : cleaning

Commit:
6.2-c : defining functions getDimensions(ctx)

Commit:
6.2-c : remove cumbersome namespace expansion

Commit:
6.2-c : simplify class VariablesBulkAddition

Commit:
6.2-c : oops, correction

Commit:
6.2c : oops, fix

Commit:
6.2-c : preparing to create and build master problem

JasonMarechal25 commented on src/modeler-optimisation-container/OptimEntityContainer.cpp:
```diff
@@ -47,12 +47,12 @@ void OptimEntityContainer::addFromSystemComponents(
         std::vector<unsigned int> modelVariableGlobalIndices;
         modelVariableGlobalIndices.resize(variables.size());
 
+        unsigned variableGlobalIndex = 0;
```
index is not global anymore.
Before : one index variable for all calls to addFromSystemComponent
After: index is reset to 0 for each call to addFromSystemComponents

Is it intended ?

Commit:
6.2-c : fix

Commit:
6.2-c : a bit of cleaning

Commit:
6.2-c : a bit of cleaning

Commit:
6.2-c : renaming

Commit:
6.2-c : creating a master and sub problem variables collector and passing it to component fillers

Commit:
[skip ci] 6.2-c : code formatting

flomnes commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -52,88 +52,123 @@ using namespace Antares::ModelerStudy::SystemModel;
+    unsigned varIndex = startIndexInPb;
+    for (const auto& name: names)
+    {
+        masterAndSubPbVars_[pbIdentifier_].push_back({name, varIndex});
```
```suggestion
        masterAndSubPbVars_[pbIdentifier_].emplace_back(name, varIndex);
```

flomnes commented on src/solver/optim-model-filler/ComponentFiller.cpp:
```diff
@@ -52,88 +52,123 @@ using namespace Antares::ModelerStudy::SystemModel;
+    return names_[index];
+}
+
+std::vector<std::string> VariableNames::names()
+{
```
copy of a possibly large object

flomnes commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -188,15 +188,27 @@ static SimplexResult OPT_TryToCallSimplex(const SingleOptimOptions& options,
+    // This instance should be created in problemeHebdo, so that it lives longer
+    // than a week, as it is currently the case.
+    std::unique_ptr<MasterAndSubPbVariables> masterAndSubPbvars; // is currently nullptr
+    if (problemeHebdo->year != 0 && optimizationNumber == PREMIERE_OPTIMISATION)
```
why compare `problemeHebdo->year` to 0 here ? Some simulations have no year 0

flomnes commented on src/modeler/modeler/Modeler.cpp:
```diff
@@ -97,85 +98,99 @@ class SystemLinearProblemBuilder final
+                                                     ortools_linear_problem,
+                                                     *data.dataSeries,
+                                                     data.scenarioGroupRepository,
+                                                     masterAndSubPbvars.get());
```
No I/O on `masterAndSubPbvars` ?

flomnes committed:
[6.2] Simplify location filtering for variable & objectives (#3200)

a-zakir commented on src/solver/optim-model-filler/include/antares/solver/optim-model-filler/ComponentFiller.h:
```diff
@@ -33,6 +33,24 @@
 namespace Antares::Optimisation
 {
 
+struct MasterAndSubPbVar
```
i'd rename this DecompositionVariables

a-zakir replied:
or ModelDecompositionVariables

a-zakir commented on src/solver/optim-model-filler/include/antares/solver/optim-model-filler/ComponentFiller.h:
```diff
@@ -33,6 +33,24 @@
+    void add(const std::vector<std::string>& names, unsigned varsCountInPb);
+
+private:
+    std::string pbIdentifier_ = "master";
```
in other places i see subproblems as default value

guilpier-code replied:
Indeed, because subproblems all have a different identifier : (year, week). So which one to choose, except "master" ?

a-zakir commented on src/modeler/modeler/Modeler.cpp:
```diff
@@ -97,85 +100,99 @@ class SystemLinearProblemBuilder final
+      0};
+
+    // Sub problem
+    auto masterAndSubPbvars = std::make_unique<MasterAndSubPbVariables>();
```
Does your decision to use a unique_ptr (heap variable) relate to the expected large size of the object?

a-zakir commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -116,24 +120,23 @@ FillContext buildFillContext(const PROBLEME_HEBDO* problemeHebdo, int NumInterva
+                        const PROBLEME_HEBDO* problemeHebdo,
+                        OptimEntityContainer& optimEntityContainer,
+                        bool namedProblems,
+                        MasterAndSubPbVariables* masterAndSubPbvars = nullptr)
```
```suggestion
void fillOrtoolsProblem(
                        FillContext& fillCtx,
                        const PROBLEME_HEBDO* problemeHebdo,
                        OptimEntityContainer& optimEntityContainer,
                        bool namedProblems,
                        MasterAndSubPbVariables* masterAndSubPbvars = nullptr)
```

guilpier-code replied:
Why asking for removing `LegacyOrtoolsLinearProblem& ortoolsProblem` from **fillOrtoolsProblem(...)** argument list ?
It seems it's used inside this function's body, is it not ?

a-zakir commented on src/solver/optimisation/opt_appel_solveur_lineaire.cpp:
```diff
@@ -116,24 +120,23 @@ FillContext buildFillContext(const PROBLEME_HEBDO* problemeHebdo, int NumInterva
     std::vector<std::unique_ptr<LinearProblemFiller>> fillersCollection;
     fillersCollection.push_back(
-      std::make_unique<LegacyFiller>(optimEntityContainer.Problem(), problemeHebdo, namedProblems));
+      std::make_unique<LegacyFiller>(ortoolsProblem, problemeHebdo, namedProblems));
```
```suggestion
      std::make_unique<LegacyFiller>(optimEntityContainer.Problem(), problemeHebdo, namedProblems));
```

a-zakir replied:
optimEntityContainer.Problem() == ortoolsProblem

flomnes commented on src/modeler/modeler/Modeler.cpp:
```diff
@@ -97,85 +100,99 @@ class SystemLinearProblemBuilder final
+      data.system->Components(),
+      [](const auto& component)
+      {
+          return std::any_of(component.getModel()->Variables().cbegin(),
```
ranges::any_of ?

flomnes reviewed (changes requested):
The logic looks good, and you made some nice simplifications to the "add variables" code.

Missing things
- Unit tests
- Actual structure.txt being written to disk

guilpier-code committed:
6.2-c : update libraries with optim config (#3199)

Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>
Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes committed:
Final steps for 6.2c (#3211)

- Fix glitches with `OptimEntityContainer`
- One `OptimEntityContainer` for each problem
- Write master.mps and 1-1.mps
- Write structure in `std::cout`

TODO
- [X] Write structure into structure.txt

---------

Co-authored-by: guilpier-code <62292552+guilpier-code@users.noreply.github.com>
Co-authored-by: guilpier <guillaume.pierre_externe@rte-france.com>
